cardano-crypto-class/src/Cardano/Crypto/SECP256K1/C.hs

```diff
@@ -12,18 +13,51 @@ module Cardano.Crypto.SECP256K1.C (
   secpSchnorrSigSignCustom,
   secpKeyPairXOnlyPub,
   secpSchnorrSigVerify,
+  secpXOnlyPubkeySerialize,
+  secpXOnlyPubkeyParse,
+  secpCtxPtr,
+  secpEcPubkeyCreate,
+  secpEcdsaSign,
+  secpEcdsaVerify,
+  secpEcCompressed,
+  secpEcPubkeySerialize,
+  secpEcdsaSignatureSerializeCompact,
+  secpEcdsaSignatureParseCompact,
+  secpEcPubkeyParse,
   ) where
 
 import Data.Bits ((.|.))
 import Foreign.Ptr (Ptr)
-import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt))
-import Cardano.Foreign
-import Cardano.Crypto.SECP256K1.Constants
+import System.IO.Unsafe (unsafePerformIO)
+import Data.Word (Word8)
+import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt), CChar)
+import Cardano.Foreign (SizedPtr (SizedPtr))
+import Cardano.Crypto.SECP256K1.Constants (
+  SECP256K1_SCHNORR_KEYPAIR_BYTES,
+  SECP256K1_SCHNORR_PRIVKEY_BYTES,
+  SECP256K1_SCHNORR_SIGNATURE_BYTES,
+  SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_SCHNORR_PUBKEY_BYTES,
+  SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_ECDSA_PRIVKEY_BYTES,
+  SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL,
+  SECP256K1_ECDSA_SIGNATURE_BYTES,
+  )
 
 data SECP256k1Context
 
 data SECP256k1SchnorrExtraParams
 
+-- We create a single context for both signing and verifying, which we use
+-- everywhere. This saves considerable time, and is safe, provided nobody
+-- outside cardano-base gets to touch it.
+--
+-- We do _not_ make this dupable, as the whole point is _not_ to compute it more
+-- than once!
+{-# NOINLINE secpCtxPtr #-}
+secpCtxPtr :: Ptr SECP256k1Context
+secpCtxPtr = unsafePerformIO . secpContextCreate $ secpContextSignVerify
+
 foreign import capi "secp256k1.h secp256k1_context_create"
```

Comment:
I thought the recommendation was to use `capi` all the time these days?

Reply:
`capi` is good for two reasons:

* It ensures we obey the platform's calling convention; and
* It checks that we don't accidentally stuff mismatched types into a C function.

The second part of this functionality [is a bit broken on our GHC](https://gitlab.haskell.org/ghc/ghc/-/issues/15531), and I'm unsure if the `SizedPtr` tricks we do help in any way, as they wrap `void*` which can be cast to absolutely anything. The main reason @lehins brought up `ccall unsafe` was probably performance: I don't know if `capi` is costlier.

Reply:
Yes, `ccall` will be faster then `capi` because `capi` will result in an extra function call and `unsafe ccall` will be faster than `safe ccall` because it doesn't need to worry of the function call being interruptible, which is totally safe in this use case

`ccall` is useful when you interface with C++ code or when you want to use CPP macros or top level constants

> it checks that we don't accidentally stuff mismatched types into a C function.

As you mentioned this is broken, and there is only very limited type matching happens anyways. You can still get all the usual integer overflows and other common bad behavior when types are mismatched, perfect example is the signed vs unsigned mismatched types in this PR, there was no protection against that

Reply:
Yeah, I think `unsafe ccall` is definitely better in this case. Plus, I learned something new today!

Reply:
I was going off https://well-typed.com/blog/2021/08/capi-usage/